### PR TITLE
JAMES-2540 Adding deque metrics and metric mailqueue contract

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            .forEach(Throwing.intConsumer(time -> mailQueue.deQueue().done(true)));
+    }
+
+    @Test
+    default void enqueueShouldIncreaseMetric() throws Exception {
+        Metric enqueuedMailsMetric = enqueuedMailsMetric();
+
+        enQueueMail(2);
+
+        verify(enqueuedMailsMetric, times(2)).increment();
```
Comment (mbaechler) [inline server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java]: how do you know enqueuedMailsMetric is a mock ?

**Comment by trantienduchn** [comment]
```
[72e77af833905dc7d38d82a26b116ab156391d25] [INFO] -------------------------------------------------------------
[72e77af833905dc7d38d82a26b116ab156391d25] [ERROR] COMPILATION ERROR : 
[72e77af833905dc7d38d82a26b116ab156391d25] [INFO] -------------------------------------------------------------
[72e77af833905dc7d38d82a26b116ab156391d25] [ERROR] /james-project/server/queue/queue-activemq/src/main/java/org/apache/james/queue/activemq/ActiveMQMailQueue.java:[108,9] constructor JMSMailQueue in class org.apache.james.queue.jms.JMSMailQueue cannot be applied to given types;
[72e77af833905dc7d38d82a26b116ab156391d25]   required: javax.jms.ConnectionFactory,org.apache.james.queue.api.MailQueueItemDecoratorFactory,java.lang.String,org.apache.james.metrics.api.MetricFactory,org.apache.james.metrics.api.GaugeRegistry
[72e77af833905dc7d38d82a26b116ab156391d25]   found: javax.jms.ConnectionFactory,org.apache.james.queue.api.MailQueueItemDecoratorFactory,java.lang.String,org.apache.james.metrics.api.MetricFactory
[72e77af833905dc7d38d82a26b116ab156391d25]   reason: actual and formal argument lists differ in length
[72e77af833905dc7d38d82a26b116ab156391d25] [ERROR] /james-project/server/queue/queue-activemq/src/main/java/org/apache/james/queue/activemq/ActiveMQMailQueueFactory.java:[39,9] constructor JMSMailQueueFactory in class org.apache.james.queue.jms.JMSMailQueueFactory cannot be applied to given types;
```

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java (hunk 1 above)
Comment (trantienduchn) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java]: At that contract level, No I don't, I created the mock at the impl level. Is this only the naming problem or some things else?

File: server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java [hunk 2]
```diff
@@ -451,6 +456,30 @@ private void setMailAttribute(Message message, Mail mail, String name) {
         }
     }
 
+    private void metricForDecrement() {
+        try {
+            dequeuedMailsMetric.decrement();
+            gaugeRegistry.register("mailQueueSize:" + queueName, queueSizeGauge());
+        } catch (Exception e) {
+            LOGGER.error("Unexpected exception while decreasing metric");
```
Comment (rouazana) [inline server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java]: you missed to log the exception

File: server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java [hunk 3]
```diff
@@ -451,6 +456,30 @@ private void setMailAttribute(Message message, Mail mail, String name) {
...
+        }
+    }
+
+    private void metricForIncrement() {
+        try {
+            enqueuedMailsMetric.increment();
+            gaugeRegistry.register("mailQueueSize:" + queueName, queueSizeGauge());
+        } catch (Exception e) {
+            LOGGER.error("Unexpected exception while increasing metric");
```
Comment (rouazana) [inline server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java]: here also

**Comment by trantienduchn** [comment]
by the way, doing mock stuffs inside `JMSMailQueueTest` may be bad, I will try to isolate mocking stuffs only in `MailQueueMetricContract` bound

File: server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java [hunk 4]
```diff
@@ -451,6 +456,30 @@ private void setMailAttribute(Message message, Mail mail, String name) {
         }
     }
 
+    private void metricForDecrement() {
+        try {
+            dequeuedMailsMetric.decrement();
+            gaugeRegistry.register("mailQueueSize:" + queueName, queueSizeGauge());
```
Comment (chibenwa) [inline server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java]: The entire concept of a gauge is to be registered once, in the constructor....

Note that this makes this method useless.

File: server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java [hunk 5]
```diff
@@ -451,6 +456,30 @@ private void setMailAttribute(Message message, Mail mail, String name) {
...
+        } catch (Exception e) {
+            LOGGER.error("Unexpected exception while decreasing metric", e);
+        }
+    }
+
+    private void metricForIncrement() {
+        try {
+            enqueuedMailsMetric.increment();
+            gaugeRegistry.register("mailQueueSize:" + queueName, queueSizeGauge());
```
Comment (chibenwa) [inline server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java]: Idem

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+import org.apache.james.queue.api.MailQueueContract;
+import org.apache.mailet.base.test.FakeMail;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+
+import com.github.fge.lambdas.Throwing;
+
+@ExtendWith(JMSMailQueueMetricExtension.class)
+public interface JMSMailQueueMetricContract extends MailQueueContract {
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: This should be global to all mail queues and not just JMS.

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        Metric enqueuedMailsMetric = testSystem.getMockEnqueuedMailsMetric();
+
+        enQueueMail(testSystem, 2);
+
+        verify(enqueuedMailsMetric, times(2)).increment();
+        verifyNoMoreInteractions(enqueuedMailsMetric);
+    }
+
+    @Test
+    default void enqueueShouldRegisterGauge(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem) throws Exception {
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: No. Creating the mailQueue should register the gauge.

And when I call the gauge, MailQueue::getSize needs to be called.

Note that implementing this test needs to SPY mailqueue and use argument collection to retrieve the gauge

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        ArgumentCaptor<Gauge> gaugeCaptor = ArgumentCaptor.forClass(Gauge.class);
+        verify(gaugeRegistry, times(2)).register(anyString(), gaugeCaptor.capture());
+        verifyNoMoreInteractions(gaugeRegistry);
+
+        assertThat(gaugeCaptor.getValue().get())
+            .isEqualTo(2L);
+    }
+
+    @Test
+    default void enqueueShouldNotDecreaseMetric(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem) throws Exception {
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: Should not decrease which metric?

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+        enQueueMail(testSystem, 1);
+
+        MailQueue.MailQueueItem mailQueueItem = testSystem.getMockMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getName())
+            .isEqualTo("name1");
+    }
+
+    @Test
+    default void dequeueShouldDecreaseMetric(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem) throws Exception {
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: which metric?

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+        enQueueMail(testSystem, 2);
+        deQueueMail(testSystem, 2);
+
+        verify(dequeuedMailsMetric, times(2)).decrement();
+        verifyNoMoreInteractions(dequeuedMailsMetric);
+    }
+
+    @Test
+    default void dequeueShouldRegisterGauge(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem) throws Exception {
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: Idem

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        ArgumentCaptor<Gauge> gaugeCaptor = ArgumentCaptor.forClass(Gauge.class);
+        verify(gaugeRegistry, times(5)).register(anyString(), gaugeCaptor.capture());
+        verifyNoMoreInteractions(gaugeRegistry);
+
+        assertThat(gaugeCaptor.getValue().get())
+            .isEqualTo(1L);
+    }
+
+    @Test
+    default void dequeueShouldNotIncreaseMetric(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem) throws Exception {
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: which metric?

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    default void deQueueMail(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem, Integer times) throws Exception {
+        IntStream
+            .rangeClosed(1, times)
+            .forEach(Throwing.intConsumer(time -> testSystem.getMockMailQueue().deQueue().done(true)));
+    }
+
+    @Test
+    default void enqueueShouldIncreaseMetric(JMSMailQueueMetricExtension.JMSMailQueueMetricTestSystem testSystem) throws Exception {
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: which metric?

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricExtension.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+    class JMSMailQueueMetricTestSystem {
+        private final JMSMailQueue mockMailQueue;
+        private final MetricFactory mockMetricFactory;
+        private final GaugeRegistry mockGaugeRegistry;
+        private final Metric mockEnqueuedMailsMetric;
+        private final Metric mockDequeuedMailsMetric;
+
+        public JMSMailQueueMetricTestSystem(JMSMailQueueTest jmsMailQueueTest) {
+            mockMetricFactory = spy(new NoopMetricFactory());
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricExtension.java]: it is not mockXXX but spyXXX

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricExtension.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+            mockGaugeRegistry = spy(new NoopGaugeRegistry());
+            mockEnqueuedMailsMetric = mock(Metric.class);
+            mockDequeuedMailsMetric = mock(Metric.class);
+
+            when(mockMetricFactory.generate(anyString())).thenReturn(mockEnqueuedMailsMetric, mockDequeuedMailsMetric);
+
+            this.mockMailQueue = spy(new JMSMailQueue(jmsMailQueueTest.connectionFactory,
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricExtension.java]: spyMailQueue

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueTest.java [hunk 15]
```diff
@@ -37,17 +36,24 @@
 import org.junit.jupiter.api.extension.ExtendWith;
 
 @ExtendWith(BrokerExtension.class)
-public class JMSMailQueueTest implements DelayedManageableMailQueueContract, PriorityManageableMailQueueContract, DelayedPriorityMailQueueContract {
+public class JMSMailQueueTest implements DelayedManageableMailQueueContract, PriorityManageableMailQueueContract, DelayedPriorityMailQueueContract,
+    JMSMailQueueMetricContract {
 
-    private JMSMailQueue mailQueue;
+    JMSMailQueue mailQueue;
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueTest.java]: What are these changes needed for?

Comment (trantienduchn) [reply server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueTest.java]: I need to access it in my test Extension

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java (hunk 1 above)
Comment (trantienduchn) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java]: i changed the variable name, if there is any other problem, please comment out :)

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java (hunk 6 above)
Comment (trantienduchn) [reply server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: I find it hard to do an abstract `MailQueueMetricContract ` because I need to recreate the new testee(for only metric testing) which we will inject our mock instance later. Do that stuff at `queue-api` level is require more ugly code like I did in the first commit. If anyone have a better implementation, I'm very happy

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java (hunk 7 above)
Comment (trantienduchn) [reply server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: There is a little tricky of `spy()` method, it cannot verify the `spider` method which will call a method reference inside. I did an alternative assertion in my `JMSMailQueueMetricContract.constructorShouldRegisterGetQueueSizeGauge`

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java (hunk 6 above)
Comment (aduprat) [reply server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: I think that adding a _GaugeRegistry getGaugeRegistry()_ method in **MailQueue** interface will solve your problem; that way you explain that your MailQueue has to support metrics.

**Comment by chibenwa** [comment]
Can't we also include time metrics to the contract?

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java (hunk 6 above)
Comment (chibenwa) [reply server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueMetricContract.java]: Or... Better than exposing MQ internals, add a `default or static` method in the test contract to retrieve it in the implementation...

File: server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java (hunk 1 above)
Comment (trantienduchn) [reply server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueMetricContract.java]: please review file changes, this commit is old!

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueTest.java [hunk 16]
```diff
@@ -37,17 +38,24 @@
 import org.junit.jupiter.api.extension.ExtendWith;
 
 @ExtendWith(BrokerExtension.class)
-public class JMSMailQueueTest implements DelayedManageableMailQueueContract, PriorityManageableMailQueueContract, DelayedPriorityMailQueueContract {
+public class JMSMailQueueTest implements DelayedManageableMailQueueContract, PriorityManageableMailQueueContract, DelayedPriorityMailQueueContract,
+    MailQueueMetricContract {
 
     private JMSMailQueue mailQueue;
+    private ActiveMQConnectionFactory connectionFactory;
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueTest.java]: Why are these fields needed? Can't it be "just" local variables, like before?

File: server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueTest.java [hunk 17]
```diff
@@ -37,17 +38,24 @@
 import org.junit.jupiter.api.extension.ExtendWith;
 
 @ExtendWith(BrokerExtension.class)
-public class JMSMailQueueTest implements DelayedManageableMailQueueContract, PriorityManageableMailQueueContract, DelayedPriorityMailQueueContract {
+public class JMSMailQueueTest implements DelayedManageableMailQueueContract, PriorityManageableMailQueueContract, DelayedPriorityMailQueueContract,
+    MailQueueMetricContract {
```
Comment (chibenwa) [inline server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueTest.java]: Can't we also pass these tests on top of ActiveMQ tests?

File: pom.xml [hunk 18]
```diff
@@ -2384,11 +2384,6 @@
                 <artifactId>junit-jupiter-engine</artifactId>
                 <version>${junit.jupiter.version}</version>
             </dependency>
-            <dependency>
```
Comment (chibenwa) [inline pom.xml]: why?

Comment (trantienduchn) [reply pom.xml]: duplicate `junit-jupiter-params` dependency

**Comment by chibenwa** [comment]
```
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] [ERROR] testServer(org.apache.james.app.spring.JamesAppSpringMainTest)  Time elapsed: 6.351 s  <<< ERROR!
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'mailqueuefactory' defined in class path resource [META-INF/spring/activemq-queue-context.xml]: Unsatisfied dependency expressed through constructor argument with index 3 of type [org.apache.james.metrics.api.GaugeRegistry]: : No qualifying bean of type [org.apache.james.metrics.api.GaugeRegistry] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type [org.apache.james.metrics.api.GaugeRegistry] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] 	at org.apache.james.app.spring.JamesAppSpringMainTest.testServer(JamesAppSpringMainTest.java:27)
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type [org.apache.james.metrics.api.GaugeRegistry] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] 	at org.apache.james.app.spring.JamesAppSpringMainTest.testServer(JamesAppSpringMainTest.java:27)
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] 
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] [INFO] 
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] [INFO] Results:
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] [INFO] 
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] [ERROR] Errors: 
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] [ERROR]   JamesAppSpringMainTest.testServer:27 » UnsatisfiedDependency Error creating be...
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] [ERROR] org.apache.james.app.spring.JamesSpringContextTest.springShouldLoadAndAddOnlyOneQuotaUpdaterListener(org.apache.james.app.spring.JamesSpringContextTest)
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] [ERROR]   Run 1: JamesSpringContextTest.setup:45 » UnsatisfiedDependency Error creating bean wi...
[ecd3bd6d3e7e15aca4ab323f22c69f70601c35e1] [ERROR]   Run 2: JamesSpringContextTest.tearDown:52 NullPointer
```

Looks like we need to inject the gaugeRegistry in Spring. Please look at `server/queue/queue-activemq/src/main/resources/META-INF/spring/activemq-queue-context.xml`

File: server/container/spring/src/main/resources/META-INF/org/apache/james/spring-server.xml [hunk 19]
```diff
@@ -308,6 +308,16 @@
 
     <bean id="metricRegistry" class="com.codahale.metrics.MetricRegistry"/>
 
+    <!--
```
Comment (chibenwa) [inline server/container/spring/src/main/resources/META-INF/org/apache/james/spring-server.xml]: IMO the huge gauge comment is not needed as gauge are parts of metrics

**Comment by trantienduchn** [comment]
test this please

File: server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java [hunk 20]
```diff
@@ -451,6 +457,10 @@ private void setMailAttribute(Message message, Mail mail, String name) {
         }
     }
 
+    private Gauge<Long> queueSizeGauge() {
+        return () -> Throwing.supplier(this::getSize).get();
```
[the remark this answers is not in the document]
Comment (trantienduchn) [reply server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java]: Is it bad? What would you propose?

[the remark this answers is not in the document]
Comment (trantienduchn) [reply server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java]: There is two ways I know, using the api or maintain the size counter variable by my self.

[the remark this answers is not in the document]
Comment (chibenwa) [reply server/queue/queue-jms/src/main/java/org/apache/james/queue/jms/JMSMailQueue.java]: Yes because mgt API is distributed aware + reboot friendly which a local counter metric is not!

**Comment by aduprat** [comment]
```
[ERROR] Errors: 
[ERROR]   ElasticSearchUidSearchOnIndexTest>UidSearchOnIndex.testSearchAtomsUS:52 » InvalidServerResponse
```
test this please

**Comment by aduprat** [comment]
merged
